### ISSUE TITLE
MachinePools cannot be upgraded atomically when used with managed providers

### ISSUE THREAD

**Issue by shyamradhakrishnan** [kubernetes-sigs/cluster-api#7248]
**What steps did you take and what happened:**
Many providers have built their managed node pool based on CPI machinepools. In most cases, customers can provide a custom vm image to be used by the node pool. for example here - https://github.com/oracle/cluster-api-provider-oci/blob/63f38727a6fba8ef78668fcb8187922ddf3cbfa8/exp/api/v1beta1/ocimanagedmachinepool_types.go#L180

The nodepool version is take from the machinepool spec like follows MachinePool.Spec.Template.Spec.Version.

This results in an inability to upgrade the machinepool(node pools) atomically. Because if we have to upgrade the kubernetes version, the version in CAPI MachinePool spec as well as the custom image infrastructure machinepool spec has to be changed and since they are in different objects, they cannot be changed atomically.

**What did you expect to happen:**

Machinepools can be upgraded tomically.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version:
- minikube/kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by sbueringer** [comment]
Hm, on KCP/MD/MS it would work like this:
An additional InfrastructureMachineTemplate would be created and then the KCP/MD/MS would be updated to the new Kubernetes version and the new InfrastructureMachineTemplate at the same time.

I assume something similar doesn't work with the MachinePool as the InfraMachinePool is representing the "machines" directly so there is no rotation?

**Comment by shyamradhakrishnan** [comment]
@sbueringer yeah, the current implementation of machinepools does not have a rotation, the machinepool references a direct infra machine pool. for example https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/cluster-template-aks.yaml#L58
and its the same for all providers I saw.

**Comment by CecileRobertMichon** [comment]
This isn't something we've encountered with AzureManagedMachinePools as there is no image reference and the image is managed by AKS, so changing the k8s version on the MachinePool is enough to trigger a k8s version + image upgrade of the managed service. However, the same think might be true with non-managed clusters that specify a custom image when using MachinePools.

@mboersma wonder if you have any thoughts on this one, and whether #4063 would improve this situation or maybe make it easier to improve in the future?

**Comment by fabriziopandini** [comment]
/triage accepted

Just for my understanding, is it possible to use the same "template rotation" approach used by other abstractions:
- create a new template/new infra machine pool object
- change MP setting both version and the pointer to the new template/new infra machine pool object (atomic operation)
- delete the old template/olde infra machine pool object

**Comment by shyamradhakrishnan** [comment]
@fabriziopandini the approach can be used but not currently as we dont use the concept of templates in machinepools, CAPI directly uses MachinePool infra objects. I think what you are suggesting is the right way forward and in sync with other places in CAPI.

**Comment by shyamradhakrishnan** [comment]
@fabriziopandini  One more advantage is machinepool support in ClusterClass can be added with your suggestion.

**Comment by rexagod** [comment]
With https://github.com/kubernetes-sigs/cluster-api/commit/8ae27c16d617af3f8fab433647f3c4afb71ee228, topology ensures eventual consistency in terms of version (and image) updates. However, this is not the case where topology is not enabled, in which case, both the `MachinePool` and the `InfraMachinePool` are not owned by one entity.

To put things in perspective and shed light on @CecileRobertMichon's comment above, I had Claude draw up the status for providers today:

```
  ┌────────────────────────────────────┬──────────────────────────────────────┬──────────────────────┬────────────────────────────────────────────────────────┐
  │                Type                │          Custom image field          │      Required?       │           Version-alone upgrade by default?            │
  ├────────────────────────────────────┼──────────────────────────────────────┼──────────────────────┼────────────────────────────────────────────────────────┤
  │ AKS AzureManagedMachinePool        │ none                                 │ —                    │ ✅ AKS selects image                                   │
  ├────────────────────────────────────┼──────────────────────────────────────┼──────────────────────┼────────────────────────────────────────────────────────┤
  │ GKE GCPManagedMachinePool          │ imageType (family, not pinned)       │ optional             │ ✅ GKE selects image                                   │
  ├────────────────────────────────────┼──────────────────────────────────────┼──────────────────────┼────────────────────────────────────────────────────────┤
  │ EKS AWSManagedMachinePool          │ amiVersion/amiType/LT ami            │ optional             │ ✅ EKS: "latest for the K8s version"                   │
  ├────────────────────────────────────┼──────────────────────────────────────┼──────────────────────┼────────────────────────────────────────────────────────┤
  │ OKE OCIManagedMachinePool          │ nodeSourceViaImage.imageId           │ optional             │ ✅ CAPOCI auto-derives -OKE-<version>-                 │
  ├────────────────────────────────────┼──────────────────────────────────────┼──────────────────────┼────────────────────────────────────────────────────────┤
  │ Azure AzureMachinePool (VMSS)      │ spec.template.image                  │ optional             │ ✅ provider defaults via GetDefaultLinuxImage(version) │
  ├────────────────────────────────────┼──────────────────────────────────────┼──────────────────────┼────────────────────────────────────────────────────────┤
  │ AWS AWSMachinePool (ASG)           │ LT ami.id                            │ optional             │ 🟡 version→AMI lookup, if a matching AMI is published  │
  ├────────────────────────────────────┼──────────────────────────────────────┼──────────────────────┼────────────────────────────────────────────────────────┤
  │ OCI OCIMachinePool (instance pool) │ instanceSourceViaImageConfig.imageId │ effectively required │ ❌ no derivation — the image is the version carrier    │
  └────────────────────────────────────┴──────────────────────────────────────┴──────────────────────┴────────────────────────────────────────────────────────┘
```


I believe the issue here is relevant for standalone environments that use custom pinned images (or require one OOTB, i.e., the last case).

I want to ensure this is something that we want to add support for before I send a patch.
